enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -180,4 +188,71 @@ impl<'a> Handler<'a> {
             ret
         }
     }
+
+    /// Do a readv() syscall
+    pub fn readv(&mut self) -> u64 {
+        let fd = self.aex.gpr.rdi;
+        let iovec = self.aex.gpr.rsi as *mut u8;
+        let iovcnt = self.aex.gpr.rdx as usize;
+
+        // Add up total size of buffers.
+        let iovec = unsafe { from_raw_parts_mut(iovec as *mut Iovec, iovcnt) }; // ???
+        let mut bufsize: usize = 0;
+        for i in 0..iovcnt {
+            bufsize += iovec[i].size;
+        }
+
+        // Allocate some unencrypted memory.
+        let size = bufsize + size_of::<Iovec>() * iovcnt;
+        let map = match self.ualloc(size as u64) {
+            Err(errno) => return errno.into_syscall(),
+            Ok(map) => map,
+        };
+
+        // Cast beginning of allocated memory as iovec slice of length iovcnt.
+        let map = unsafe { from_raw_parts_mut(map as *mut Iovec, iovcnt) };
+
+        // Set pointers in unencrypted iovec slice to use the rest of the allocated memory.
+        let mut chunk = map.as_ptr() as usize + size_of::<Iovec>() * iovcnt; // ??? May be unsafe or illegal as per from_raw_parts_mut rules
+        for i in 0..iovcnt {
+            if chunk > map.as_ptr() as usize + size {
+                // ???
+                // TODO: What error to return?
+                self.attacked();
+            }
+            map[i].base = chunk as *mut u8;
+            chunk = chunk + iovec[i].size;
+        }
+
+        unsafe {
+            // Do the syscall; replace encrypted memory with unencrypted memory.
+            let ret = self.syscall(
+                SysCall::READV,
+                fd,
+                map.as_ptr() as _,
+                iovcnt as u64,
+                0,
+                0,
+                0,
+            ); // ???
+            self.ufree(map.as_ptr() as *mut u8, size as u64); // ???
+
+            // Copy the unencrypted input into encrypted memory.
+            if ErrNo::from_syscall(ret).is_none() {
+                if ret > size as u64 {
+                    self.attacked();
+                }
+
+                for i in 0..iovcnt {
+                    core::ptr::copy_nonoverlapping(
+                        map[i].base as *mut u8,
+                        iovec[i].base as *mut u8,
+                        ret as _,
+                    );
+                }
```

Comment:
Iterate over the slice rather than using an index. However, keep in mind that we handed the whole array of iovecs to the untrusted host. That means that we need to validate that the values haven't been manipulated. For example, the host might replace `iovec[0].base` with a pointer to enclave memory in order to trick us into copying encrypted memory somewhere else. So, like we did when initializing the array, we need to iterate over both slices. Validate that the untrusted iovec hasn't been modified (and trigger `attacked()` if it has; that is suspicious behavior). Then do the copy.

Alternatively, we could take advantage of the contiguousness of the `map` buffer to bypass the `untrusted` array altogether. This would probably be fewer instructions. However, we'd bypass an opportunity to detect malicious behavior. So I'd prefer to be slightly slower and more security conscious.

Reply:
> For example, the host might replace iovec[0].base with a pointer to enclave memory ... Validate that the untrusted iovec hasn't been modified

@npmccallum For now, I'm assuming this validation means to check that none of the `base` pointers refer to something inside enclave memory. If there are other checks that should be done here, please let me know. I'll also continue to think on it.

I'm fine with doing the slower and more security-conscious version.

Reply:
Testing that the base pointers (and base+size pointers) aren't inside the enclave is a weaker version of the testing I had in mind. I wanted to recreate the exact expected `base` and `size` that it should be and validate that the results haven't been tampered with in any way.